File: DesignPattenStudy/DecoratorPattern/Java/Concrete/MilkDecorator.java
package DecoratorPattern.Java.Concrete;

import DecoratorPattern.Java.Abstract.Component;
import DecoratorPattern.Java.Abstract.Decorator;

public class MilkDecorator extends Decorator {

    // coffee latte
    public MilkDecorator(Component component) {

        super(component);
    }

    @Override
    public Integer addPrice() {

        return super.addPrice() + new EspressoDecorator(new BaseComponent()).addPrice() + 2000;
    }

    @Override
    public String showItem() {
        return super.showItem() + " + Milk : CoffeeLatte";
    }
}

File: DesignPattenStudy/DecoratorPattern/Kotlin/Concrete/MilkDecorator.kt
package DecoratorPattern.Kotlin.Concrete

import DecoratorPattern.Kotlin.Astract.Component
import DecoratorPattern.Kotlin.Astract.Decorator

class MilkDecorator(override val component: Component) : Decorator(component) {

    override fun addPrice(): Int = super.addPrice() + EspressoDecorator(BaseComponent()).addPrice() + 2000

    override fun showItem(): String = "${super.showItem()} + Milk : CoffeeLatte"
}